File: app — копия (4)/src/main/java/com/dnd/app/ui/screens/character_creator/tabs/RaceStep.kt
// Имя файла: app/src/main/java/com/dnd/app/ui/screens/character_creator/tabs/RaceStep.kt
// --- НАЧАЛО ФАЙЛА ---
package com.dnd.app.ui.screens.character_creator.tabs

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.sp
import com.dnd.app.domain.model.ChoiceOption
import com.dnd.app.domain.model.ChoiceResult
import com.dnd.app.domain.model.Feature
import com.dnd.app.domain.model.Race
import com.dnd.app.ui.screens.character_creator.components.* // Импортирует WizardUiConfig и компоненты
import com.dnd.app.util.stripHtml

@Composable
fun RaceStep(
    availableRaces: List<Race>,
    selectedRaceIndex: String,
    onRaceSelect: (String) -> Unit,
    availableSubraces: List<Race>,
    selectedSubraceIndex: String?,
    onSubraceSelect: (String) -> Unit,
    baseFeatures: List<Feature>,
    subraceFeatures: List<Feature>,
    currentSelections: Map<String, ChoiceResult>,
    onSelectionChanged: (String, ChoiceResult) -> Unit,
    globalExclusions: Set<String>,
    selectedFeatDetails: Feature?,
    expandedStates: MutableMap<String, Boolean>
) {
    LazyColumn(
        modifier = Modifier.fillMaxSize().padding(WizardUiConfig.WIZARD_STEP_SCREEN_PADDING),
        verticalArrangement = Arrangement.spacedBy(WizardUiConfig.WIZARD_STEP_SECTION_GAP)
    ) {
        item {
            FlatWizardSection(title = "Раса") {
                val opts = availableRaces.map { ChoiceOption(it.index, it.name) }
                SmartDropdown(opts, selectedRaceIndex, onSelected = { onRaceSelect(it.id) })
            }
        }

        items(baseFeatures, key = { "base_f_${it.id}_${it.index}" }) { feature ->
            if (feature.isSubraceSelector) {
                FlatWizardSection(title = feature.name) {
                    val opts = availableSubraces.map { ChoiceOption(it.index, it.name) }
                    SmartDropdown(
                        opts,
                        selectedSubraceIndex,
                        onSelected = { onSubraceSelect(it.id) },
                        placeholder = "Выберите разновидность..."
                    )
                }
            } else {
                // [ИЗМЕНЕНО] Логика выбора черты теперь применяется и к базовым способностям расы.
                if (feature.index.contains("feat")) {
                    FlatWizardSection(title = feature.name) {
                        FeatSelectionCard(
                            grantingFeature = feature,
                            selectedFeatDetails = selectedFeatDetails,
                            allSelections = currentSelections,
                            proficiencyExclusions = globalExclusions,
                            onSelectionUpdated = onSelectionChanged,
                            expandedStates = expandedStates
                        )
                    }
                } else {
                    FeatureSection(
                        feature = feature,
                        allSelections = currentSelections,
                        onSelectionChanged = onSelectionChanged,
                        globalExclusions = globalExclusions,
                        expandedStates = expandedStates
                    )
                }
            }
        }

        items(subraceFeatures, key = { "sub_f_${it.id}_${it.index}" }) { feature ->
            if (feature.index.contains("feat")) {
                FlatWizardSection(title = feature.name) {
                    FeatSelectionCard(
                        grantingFeature = feature,
                        selectedFeatDetails = selectedFeatDetails,
                        allSelections = currentSelections,
                        proficiencyExclusions = globalExclusions,
                        onSelectionUpdated = onSelectionChanged,
                        expandedStates = expandedStates
                    )
                }
            } else {
                FeatureSection(
                    feature = feature,
                    allSelections = currentSelections,
                    onSelectionChanged = onSelectionChanged,
                    globalExclusions = globalExclusions,
                    expandedStates = expandedStates
                )
            }
        }
    }
}

@Composable
private fun FeatureSection(
    feature: Feature,
    allSelections: Map<String, ChoiceResult>,
    onSelectionChanged: (String, ChoiceResult) -> Unit,
    globalExclusions: Set<String>,
    expandedStates: MutableMap<String, Boolean>
) {
    if (feature.name == "Описание" && feature.description.isBlank()) return

    FlatWizardSection(title = feature.name) {
        Column(verticalArrangement = Arrangement.spacedBy(WizardUiConfig.WIZARD_STEP_SECTION_GAP)) {
            if (feature.description.isNotBlank()) {
                Text(
                    text = feature.description.stripHtml(),
                    fontSize = WizardUiConfig.FONT_SIZE_CONTENT,
                    lineHeight = 18.sp
                )
            }
            if (feature.embeddedSpells.isNotEmpty()) {
                val key = "spells_${feature.index}"
                val isExpanded = expandedStates.getOrPut(key) { true }
                CollapsibleChoiceSection(
                    title = "Добавленные заклинания",
                    expanded = isExpanded,
                    onToggle = { expandedStates[key] = !isExpanded }
                ) {
                    Column {
                        feature.embeddedSpells.forEach { spell ->
                            val spellExpandedKey = "granted_${feature.index}_${spell.index}"
                            val isSpellExpanded = expandedStates.getOrPut(spellExpandedKey) { false }
                            UnifiedSpellListItem(
                                spell = spell,
                                actionType = SpellActionType.ADD,
                                onActionClick = {},
                                isActionEnabled = false,
                                isExpanded = isSpellExpanded,
                                onToggleExpand = { expandedStates[spellExpandedKey] = !isSpellExpanded }
                            )
                        }
                    }
                }
            }
            feature.choices.forEach { choice ->
                FeatureChoiceBlock(
                    choice = choice,
                    allSelections = allSelections,
                    onSelectionUpdated = onSelectionChanged,
                    selectionKey = feature.index,
                    expandedStates = expandedStates,
                    globalExclusions = globalExclusions
                )
            }
        }
    }
}
// --- КОНЕЦ ФАЙЛА ---
// Имя файла: app/src/main/java/com/dnd/app/ui/screens/character_creator/tabs/RaceStep.kt

File: app — копия (2)/src/main/java/com/dnd/app/ui/screens/character_creator/tabs/RaceStep.py
// Имя файла: app/src/main/java/com/dnd/app/ui/screens/character_creator/tabs/RaceStep.kt
// --- НАЧАЛО ФАЙЛА ---
package com.dnd.app.ui.screens.character_creator.tabs

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.dnd.app.domain.model.*
import com.dnd.app.ui.screens.character_creator.components.*
import com.dnd.app.util.stripHtml

@Composable
fun RaceStep(
    availableRaces: List<Race>,
    selectedRaceIndex: String,
    onRaceSelect: (String) -> Unit,
    availableSubraces: List<Race>,
    selectedSubraceIndex: String?,
    onSubraceSelect: (String) -> Unit,
    features: List<Feature>,
    currentSelections: Map<String, ChoiceResult>,
    onSelectionChanged: (String, ChoiceResult) -> Unit,
    globalExclusions: Set<String>
) {
    // Сохраняем состояние скролла между рекомпозициями
    val listState = rememberLazyListState()

    LazyColumn(
        state = listState,
        modifier = Modifier.fillMaxSize().padding(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        // 1. СЕЛЕКТОР ОСНОВНОЙ РАСЫ
        item(key = "main_race_dropdown_selector") {
            FlatWizardSection(title = "Раса") {
                val options = availableRaces.map { ChoiceOption(it.index, it.name) }
                SmartDropdown(options, selectedRaceIndex, onSelected = { onRaceSelect(it.id) })
            }
        }

        // 2. ДИНАМИЧЕСКИЙ СПИСОК ФИЧ
        // Ключ включает ID фичи, что гарантирует стабильность позиции существующих элементов
        items(features, key = { "feat_${it.index}_${it.id}" }) { feature ->
            if (feature.isSubraceSelector) {
                // СЕЛЕКТОР ПОДРАСЫ: Появляется ровно там, где его поставил репозиторий
                FlatWizardSection(title = feature.name) {
                    Column {
                        if (feature.description.isNotBlank()) {
                            Text(
                                text = feature.description,
                                fontSize = 13.sp,
                                color = Color.DarkGray,
                                modifier = Modifier.padding(bottom = 8.dp)
                            )
                        }
                        val options = availableSubraces.map { ChoiceOption(it.index, it.name) }
                        SmartDropdown(
                            options = options,
                            selectedId = selectedSubraceIndex,
                            onSelected = { onSubraceSelect(it.id) },
                            placeholder = "Выберите разновидность..."
                        )
                    }
                }
            } else {
                // ОБЫЧНАЯ ФИЧА
                FlatWizardSection(title = feature.name) {
                    Column {
                        if (feature.description.isNotBlank()) {
                            Text(
                                text = feature.description.stripHtml(),
                                fontSize = 14.sp,
                                color = Color.Black,
                                lineHeight = 17.sp,
                                modifier = Modifier.padding(bottom = if (feature.choices.isNotEmpty() || feature.embeddedSpells.isNotEmpty()) 8.dp else 0.dp)
                            )
                        }

                        feature.embeddedSpells.forEach { EmbeddedSpellRow(it) }

                        feature.choices.forEach { choice ->
                            FeatureChoiceBlock(
                                choice = choice,
                                currentSelection = currentSelections[feature.index],
                                onSelectionChanged = { res -> onSelectionChanged(feature.index, res) },
                                globalExclusions = globalExclusions
                            )
                        }
                    }
                }
            }
        }

        item(key = "race_bottom_padding") { Spacer(Modifier.height(32.dp)) }
    }
}
// --- КОНЕЦ ФАЙЛА ---
// Имя файла: app/src/main/java/com/dnd/app/ui/screens/character_creator/tabs/RaceStep.kt